libraries/WiFi/src/WiFi.cpp

```diff
@@ -1,99 +1,217 @@
 #include "WiFi.h"
 
-WiFiClass WiFi;
+// Static Wi-Fi state instance
+struct WiFiState {
+	struct net_if *sta_iface = nullptr;
+	struct net_if *ap_iface = nullptr;
+	struct wifi_connect_req_params ap_config;
+	struct wifi_connect_req_params sta_config;
+	struct wifi_iface_status sta_state = {0};
+	struct wifi_scan_result scanResults[MAX_SCAN_RESULTS];
+	uint8_t resultCount = 0;
+	struct net_mgmt_event_callback wifiCb;
+	bool soughtNetworkFound = false;
+	bool scanSequenceFinished = false;
+};
 
-String WiFiClass::firmwareVersion() {
-#if defined(ARDUINO_PORTENTA_C33)
-	return "v1.5.0";
-#else
-	return "v0.0.0";
-#endif
+WiFiClass::WiFiClass() {
 }
 
+WiFiClass::~WiFiClass() {
+}
+
+// Static instance of Wi-Fi state
+struct WiFiState WiFiClass::wifiState;
+
 int WiFiClass::begin(const char *ssid, const char *passphrase, wl_enc_type security,
 					 bool blocking) {
-	sta_iface = net_if_get_wifi_sta();
-	netif = sta_iface;
-	sta_config.ssid = (const uint8_t *)ssid;
-	sta_config.ssid_length = strlen(ssid);
-	sta_config.psk = (const uint8_t *)passphrase;
-	sta_config.psk_length = strlen(passphrase);
-	// TODO: change these fields with scan() results
-	sta_config.security = WIFI_SECURITY_TYPE_PSK;
-	sta_config.channel = WIFI_CHANNEL_ANY;
-	sta_config.band = WIFI_FREQ_BAND_2_4_GHZ;
-	sta_config.bandwidth = WIFI_FREQ_BANDWIDTH_20MHZ;
-	int ret = net_mgmt(NET_REQUEST_WIFI_CONNECT, sta_iface, &sta_config,
-					   sizeof(struct wifi_connect_req_params));
-	if (ret) {
-		return false;
+	wifi_security_type wifi_security = convert_enc_type_to_security_type(security);
+
+	wifiState.sta_iface = net_if_get_wifi_sta();
+	netif = wifiState.sta_iface;
+	wifiState.sta_config.ssid = (const uint8_t *)ssid;
+	wifiState.sta_config.ssid_length = strlen(ssid);
+	wifiState.sta_config.psk = (const uint8_t *)passphrase;
+	wifiState.sta_config.psk_length = strlen(passphrase);
+
+	// Set Wi-Fi security type if specified
+	if (wifi_security != WIFI_SECURITY_TYPE_NONE) {
+		wifiState.sta_config.security = wifi_security;
+	} else {
+		wifiState.sta_config.security = WIFI_SECURITY_TYPE_PSK;
 	}
 
-	ret = status();
-	if (ret != WL_CONNECTED && blocking) {
-		net_mgmt_event_wait_on_iface(sta_iface, NET_EVENT_WIFI_CONNECT_RESULT, NULL, NULL, NULL,
-									 K_FOREVER);
+	wifiState.sta_config.channel = WIFI_CHANNEL_ANY;
+	wifiState.sta_config.band = WIFI_FREQ_BAND_2_4_GHZ;
+	wifiState.sta_config.bandwidth = WIFI_FREQ_BANDWIDTH_20MHZ;
+
+	// Register the Wi-Fi event callback
+	net_mgmt_init_event_callback(&wifiState.wifiCb, scanEventDispatcher,
+								 NET_EVENT_WIFI_SCAN_RESULT | NET_EVENT_WIFI_SCAN_DONE);
+	net_mgmt_add_event_callback(&wifiState.wifiCb);
```

Comment:
Does his operation remove the `NetworkInterface` event callback? In that case we have to provide an api on NetworkInterface to extend the callback capabilities